### PR TITLE
[LLVM] [Clang] Support for Gentoo `*t64` triples (64-bit time_t ABIs)

### PR DESCRIPTION
**Pull request by mgorny**
Gentoo is planning to introduce a `*t64` suffix for triples that will be used by 32-bit platforms that use 64-bit `time_t`. Add support for parsing and accepting these triples, and while at it make clang automatically enable the necessary glibc feature macros when this suffix is used.

An open question is whether we can backport this to LLVM 19.x. After all, adding new triplets to Triple sounds like an ABI change — though I suppose we can minimize the risk of breaking something if we move new enum values to the very end.

### CODE REVIEW

File: clang/test/Preprocessor/time64.c [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// RUN: %clang_cc1 -E -dM -triple=i686-pc-linux-gnu < /dev/null | FileCheck -match-full-lines -check-prefix TIME32 %s
```
Comment (MaskRay) [inline clang/test/Preprocessor/time64.c:1]: You can omit `< `

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -610,7 +610,8 @@ static llvm::Triple computeTargetTriple(const Driver &D,
     if (A->getOption().matches(options::OPT_m64) ||
         A->getOption().matches(options::OPT_maix64)) {
       AT = Target.get64BitArchVariant().getArch();
-      if (Target.getEnvironment() == llvm::Triple::GNUX32)
+      if (Target.getEnvironment() == llvm::Triple::GNUX32 ||
```
Comment (MaskRay) [inline clang/lib/Driver/Driver.cpp:613]: Optional: you could simplify the pattern  with `llvm::is_contained({...}, X)`

**Review by MaskRay — APPROVED** [review]

File: clang/test/Preprocessor/time64.c (hunk 1 above)
Comment (mgorny) [reply clang/test/Preprocessor/time64.c:1]: Yeah, I was wondering if it's in the `init.c` tests for a reason.

File: llvm/unittests/TargetParser/TripleTest.cpp [hunk 3]
```diff
@@ -1283,6 +1283,28 @@ TEST(TripleTest, ParsedIDs) {
...
+  EXPECT_EQ(Triple::Linux, T.getOS());
+  EXPECT_EQ(Triple::GNUEABIT64, T.getEnvironment());
+  EXPECT_TRUE(T.isTime64ABI());
+
+  T = Triple("armv7a-unknown-linux-gnueabihft64");
+  EXPECT_EQ(Triple::arm, T.getArch());
+  EXPECT_EQ(Triple::UnknownVendor, T.getVendor());
+  EXPECT_EQ(Triple::Linux, T.getOS());
+  EXPECT_EQ(Triple::GNUEABIHFT64, T.getEnvironment());
+  EXPECT_TRUE(T.isTime64ABI());
```
Comment (MaskRay) [inline llvm/unittests/TargetParser/TripleTest.cpp:1306]: needs a negative test for `isTime64ABI`

Comment (mgorny) [reply llvm/unittests/TargetParser/TripleTest.cpp:1315]: Added.

File: clang/lib/Driver/Driver.cpp (hunk 2 above)
Comment (mgorny) [reply clang/lib/Driver/Driver.cpp:613]: To be honest, I don't have the energy to rewrite it all manually, especially that these files don't seem to be clang-formatted consistently.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 4]
```diff
@@ -177,6 +177,7 @@ createTargetCodeGenInfo(CodeGenModule &CGM) {
     else if (CodeGenOpts.FloatABI == "hard" ||
              (CodeGenOpts.FloatABI != "soft" &&
               (Triple.getEnvironment() == llvm::Triple::GNUEABIHF ||
```
Comment (mgorny) [inline clang/lib/CodeGen/CodeGenModule.cpp:179]: I'm wondering if we should add a `isHardFloatABI()` or something like that to the `Triple` class.

Comment (MaskRay) [reply clang/lib/CodeGen/CodeGenModule.cpp:179]: This helper sounds good.

will we have `muslt32`?

Comment (mgorny) [reply clang/lib/CodeGen/CodeGenModule.cpp:179]: Are you asking if we're going to have musl suffixes on Gentoo? No, musl already switched to 64-bit `time_t`.

**Review by thesamesam — APPROVED** [review]

**Pull request merged**